i.MX RT685's HiFi 4 DSP support

VitekST opened the pull request:
This PR introduces support for the Xtensa LX6 augmented with the HiFi 4 DSP extension on the i.MX RT685, this time compatible with HWMv2. 

It instantiates drivers for basic peripherals (GPIO, Flexcomm UART), as well as peripherals responsible for audio (Flexcomm I2S, DMA). Audio playback and capture is enabled and functional. Hardware initialisation, code loading and run control from the CM33 side is also implemented and the support is backed by a set of samples.

VitekST committed:
dts: xtensa: Add DT for HiFi 4 on NXP i.MX RT685

Add a basic device tree for the HiFi 4 DSP on the NXP i.MX RT685
microcontroller.

While no peripherals are instantiated, Zephyr is able to technically run
on this target.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
boards: mimxrt685_evk/mimxrt685s/hifi4: Enable GPIO and UART

Modify HiFi 4 domain's DT to instantiate GPIO peripherals and the 0th
Flexcomm peripheral in UART mode. Also instantiate associated
peripherals - Pinctrl, Clkctl, PINT.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
boards: xtensa: Add mimxrt685_evk/mimxrt685s/hifi4

Add board definitions for the HiFi 4 DSP located on the MIMXRT685-EVK.
Also add a basic board-level DT to get Zephyr to technically run.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
soc: mimxrt685s: Enable MU

Instantiate the MU driver (mbox_nxp_imx_mu.c) for both the CM33 domain
and the HiFi4 domain of the NXP i.MX RT685. Add neccessary definitions
for instantiated boards.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
drivers: intc_nxp_pint: Decouple from fsl_power.h

Add drivers/interrupt_controller/intc_nxp_pint/power.h abstracting
EnableDeepSleepIRQ and DisableDeepSleepIRQ invocations from
intc_nxp_pint.c. Modify intc_nxp_pint.c to use that file.

fsl_power.c and fsl_power.h can't be built on the
mimxrt685/mimxrt685s/hifi4 target, so it's excluded from it in hal_nxp.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
drivers: misc: Add nxp,rtxxx-dsp-ctrl

Add the nxp,rtxxx-dsp-ctrl driver.

Responsibility of this driver is to load code executed by Xtensa-family
cores on NXP i.MX RTxxx microcontrollers and to control their run.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
dts: mimxrt685s/cm33: Instantiate nxp,rtxxx-dsp-ctrl

Instantiate the nxp,rtxxx-dsp-ctrl driver to enable DSP code upload and
run control.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
soc: mimxrt685s/hifi4: Add init code

Add init code - responsible for clock attachment of peripherals
responsible for audio playback (DMA, Flexcomms in I2S mode, ...) and for
INPUTMUX settings (IRQ line routing).

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
dts: mimxrt685s/hifi4: Instantiate devs for audio

Instantiate Flexcomm #1, Flexcomm #3, DMA #1 and I3C - these are
peripherals responsible for audio playback and capture on the i.MX
RT685's HiFi 4 domain.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
board: mimxrt685_evk/mimxrt685s/hifi4: Add init code

Add initialisation code for the HiFi 4 DSP domain of the mimxrt685_evk
board. Responsible for setting up signal sharing between Flexcomm #1 and
Flexcomm #3 peripherals both in the I2S mode (simultaneous audio
playback and capture).

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
dts: mimxrt685_evk/mimxrt685s/hifi4: Enable devs for audio

Instantiate, configure and enable Flexcomm #1, Flexcomm #3, DMA, I3C for
audio capture and playback on mimxrt685_evk's HiFi 4 DSP domain. Add
pinmux definitions for those.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

dbaluta commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -825,11 +829,35 @@ static int dma_mcux_lpc_get_status(const struct device *dev, uint32_t channel,
+		case DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT:
+			*value = 4;
+			return 0;
+
```
I think you should go with `break` instead of `return 0` and add the return 0 at the bottom of the function.

dbaluta commented on dts/arm/nxp/nxp_rt6xx_common.dtsi:
```diff
@@ -60,9 +60,14 @@
 				reg = <0x24000000 DT_SIZE_K(64)>;
 				type = <NXP_RTXXX_ADSP_REGION_DATA>;
+			};*/
+
```
We shouldn't have commented code in the final version. So you should remove this. One can look at the history and see the changes.

dbaluta commented on drivers/audio/wm8904.c:
```diff
@@ -264,17 +264,17 @@ static int wm8904_in_update(
 static int wm8904_in_volume_config(const struct device *dev, audio_channel_t channel, int volume)
 {
```
This is a bugfix unrelated to the original goal of this PR. Please create another PR for it. Also please fix warnings like this:

![image](https://github.com/user-attachments/assets/40d187d6-6be9-4224-8793-21ccdeed01ba)

dbaluta commented:
@VitekST excellent work. Since this PR is still in draft and depending on other PR  I think we can create separate PR for unrelated code here:

So, we can move in seperate PRs following patches:

1) [dma_mcux_lpc: add host directions, DMA driver ctx, getter func](https://github.com/zephyrproject-rtos/zephyr/pull/77814/commits/1167a7d0a44183e58c7c40a2f71955c46343d971)
2) [drivers: wm8904: Fix audio input configuration](https://github.com/zephyrproject-rtos/zephyr/pull/77814/commits/756aeb0d55311feab6c560d58cb863d5c9f35d64)
3) Maybe some more which I cannot figure out right now.

A PR with more than 7-8 patches will be very hard to review and get traction from community. Lets put the patches PR by PR :) starting from simple ones.

VitekST committed:
manifest: Point hal_nxp to associated PR

Support for the i.MX RT685's HiFi 4 DSP domain required changes in the
hal_nxp repo, thus point it to to the HEAD of PR bearing the required
changes.

Temporary.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
soc: xtensa: Add HiFi4 DSP of NXP i.MX RT685

Add SoC definitions and a linker file for the HiFi 4 DSP on the NXP i.MX
RT685 microcontroller.

Also modify the West manifest - support depends on a different branch of
the Zephyr NXP HAL, which includes some files that were missing when
compiling for the DSP and omits some files that can't compile for the
DSP.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
modules: hal_nxp: Edit lists for mimxrt685s/hifi4

Exclude fsl_power.c from DSP domains, as it doesn't compile on those
because of missing register definitions. Include fsl_dsp.c wherever it
can be included, as it's required for initialisation of DSP domains.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
samples: Add RTxxx AMP samples

Add common structure for RTxxx AMP samples. Add amp_blink, amp_mbox,
amp_audio_output and amp_audio_loopback samples.

These samples are now configured only for the mimxrt685_evk, but are
intended to support other NXP microcontrollers with HiFi x DSP cores
at a later date.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

dleach02 commented:
@VitekST  What is the plan on this PR?

VitekST commented:
@dleach02 The plan is to rebase, resolve conflicts, resolve any points of discussion and wait for it to be merged. Its contents is still relevant to the objective I am trying to achieve.

maxekman commented:
I would also like to see this merged, it is highly relevant and the HiFi core is used in other products too.

VitekST commented:
@dleach02 @mmahadevan108 Rebased, test do pass and based on my manual testing, the support seems to work. Now I am waiting for your input.

dleach02 commented:
okay. we will sequence merging of the HAL component

iuliana-prodan commented on soc/nxp/imxrt/imxrt6xx/Kconfig.defconfig:
```diff
@@ -65,3 +65,16 @@ config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 198000000
+
+config MCUX_CORE_SUFFIX
+	default "_dsp"
```
we should use everywhere hifi4

VitekST replied:
@iuliana-prodan I agree, however, this name is tied to the MCUX SDK, from which the NXP's HAL is derived. Renaming the core suffix to something more appropriate, like `hifi4`, would imply conflicts while syncing the HAL with the release MCUX SDK, necessitating the act of renaming agood number of directories, files and in-code references to that in the SDK.

iuliana-prodan replied:
Ok, fair enough - I remember we have some restriction here.
But in other places, where we don't have a dependency on hal you can use hifi4 - like [here ](https://github.com/zephyrproject-rtos/zephyr/pull/77814/files#diff-93553fe313dd0d9e0dc8cd7b516bf36046207f3a2bf1c84b48c8210f0ff0c8e0R53)or other places like this.

iuliana-prodan commented on include/zephyr/drivers/misc/nxp_rtxxx_adsp_ctrl/nxp_rtxxx_adsp_ctrl.h:
```diff
@@ -0,0 +1,79 @@
+#include <zephyr/device.h>
+#include <zephyr/dt-bindings/misc/nxp_rtxxx_adsp_ctrl.h>
+
+#ifndef __NXP_RTXXX_ADSP_CTRL_H__
```
Here, in this file (also the naming) can it be applied for other DSPs (from other i.MXRTs) or is just for rt600?
If it applies to other you should name it DSP, not ADSP (not just for audio).

As a summary, where possible name it hifi4, if applies to multiple i.MXRTs put dsp. :)

PS: Please split this in multiple PRs, is very hard to review all these together.

Thanks!

VitekST replied:
@iuliana-prodan Yes, I intend to support other 3-digit RT devices.
As was communicated internally, I intend to split the PR.

VitekST commented:
Very well, I've split the PR: https://github.com/zephyrproject-rtos/zephyr/pull/89404
@dleach02 @iuliana-prodan @dbaluta I hope that this act gets things moving.
